File: Sorting/BreadthFirstSearch.java
import java.util.*;
class Program{

  static class Node{
    String name;
    ArrayList<Node> children = new ArrayList<Node>();

    public Node(String name)
    {
      this.name = name;
    }

  public Node addChild(String name)
  {
    Node child = new Node(name);
    children.add(child);
    return this;
  }

// O(v + e) time | O(v) space 
  public ArrayList<String> breadthFirstSearch(ArrayList<String> array)
  {
    ArrayDeque<Node> queue = new ArrayDeque<Node>();
    queue.add(this);
    while(!queue.isEmpty())
    {
      Node currentNode = queue.poll();
      array.add(currentNode.name);
      for(int i = 0; i < currentNode.children.size(); i++)
      {
        queue.add(currentNode.children.get(i));
      }
    }
    return array;
  }
}
}

File: DailyCodingProblem/C++/BreadthFirstSearch.cpp
#include <vector>
#include <deque>
using namespace std;

class Node {
  public:
    string name;
    vector<Node*> children;

    Node(string str) {
      name = str;
    }

    // O(v + e) time | O(v) space
    vector<string> breadthFirstSearch(vector<string>* array) {
			deque<Node*> queue {this};
			int counter = 0;
			while(!queue.empty())
			{
				Node current = *queue.front();
				queue.pop_front();
				array->push_back(current.name);
				for(int i = 0; i < current.children.size(); i++)
				{
					queue.push_back(current.children[i]);
				}
			}
			return *array;
    }

    Node* addChild(string name) {
      Node* child = new Node(name);
      children.push_back(child);
      return this;
    }
};
